### PR TITLE
[Ttahub 935] template race condition

### PR DESCRIPTION
**Pull request by GarrettEHill**
## Description of change
A race condition was identified where if multiple goals or objectives are triggered to create a template at the same time where they should share a template, they instead would create distinct templates.

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-935

### Every PR

- [ ] Meets issue criteria
- [ ] JIRA ticket status updated
- [ ] Code is meaningfully tested
- [ ] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [ ] Update JIRA ticket status

### CODE REVIEW

**Comment by thewatermethod** [comment]
How would I recreate the issue? Is it as simple as creating multiple goals with the same text at once?

**Comment by GarrettEHill** [comment]
> How would I recreate the issue? Is it as simple as creating multiple goals with the same text at once?

Yes, so same new goal or objective  on multiple grants then submit AR for approval.

**Comment by thewatermethod** [comment]
I created a goals with different grants and identical text. Confirmed that two goals with separate goalTemplateIds (likewise, objectives with separate objectiveTemplateIds were created). Ran your migration, and the issue was corrected for the existing goals - the goalTemplateIds were consolidated.

Went on to create a new report using the same process as before, created a report with two grants and a goal + objective. Two goals and objectives were created, as before, however in checking the database they all have "null" for the templateId column. Is this expected?

File: src/models/hooks/activityReport.js [hunk 1]
```diff
@@ -45,16 +45,23 @@ const propogateSubmissionStatus = async (sequelize, instance, options) => {
...
-        sequelize,
-        options.transaction,
-        instance.regionId,
-        goal.name,
-        goal.createdAt,
-        goal.updatedAt,
-      )));
-      await Promise.all(goals.map(async (goal, i) => sequelize.models.Goal.update(
-        { goalTemplateId: templateIds[i] },
+      const distinctGoals = [...new Map(goals.map((goal) => [goal.name, goal])).values()];
```
Comment (thewatermethod) [inline src/models/hooks/activityReport.js:48]: Would love to see some comments here, even like "create a distinct list of goals based on name".

I'm also wondering if "propogateSubmissionStatus" is kind of poor name for this function or at the least, if the function is getting kind of overloaded

**Comment by GarrettEHill** [comment]
> 

It is working for me, can you show me what your seeing?

**Comment by thewatermethod** [comment]
Sure, ping me anytime today and I can screenshare

**Comment by kryswisnaskas** [comment]
At first glance, for an AR with multiple recipients, only one template is created for a goal and only one template for an objective 👍 Nice! (this wasn't the case before for multiple recipients)
Btw, I see the templates for goals and objectives being created once the report is submitted. Assuming that's intended.

Also seeing what Matt mentioned above - the goals/objectives rows point to null templates.

**Review by hardwarehuman — APPROVED** [review]

File: src/models/tests/activityReport.test.js [hunk 2]
```diff
@@ -50,8 +50,21 @@ const mockOtherEntity = {
...
+  number: '99CH9998',
+  regionId: 2,
+  status: 'Active',
+  startDate: new Date('2021-02-09T15:13:00.000Z'),
+  endDate: new Date('2021-02-09T15:13:00.000Z'),
+  cdi: false,
+  grantSpecialistName: null,
+  grantSpecialistEmail: null,
+  stateCode: 'NY',
+  anualFundingMonth: 'October',
```
Comment (hardwarehuman) [inline src/models/tests/activityReport.test.js:64]: there's two 'n's in "annual", and it is `annualFundingMonth` in the db.

**Review by hardwarehuman — APPROVED** [review]

**Review by thewatermethod — APPROVED** [review]

**Pull request merged**